packages/core/strapi/lib/types/core/attributes/relation.d.ts

```diff
@@ -1,48 +1,62 @@
-import { SchemaUID } from '../../utils';
-import { Attribute, ConfigurableOption, PrivateOption } from './base';
-import { GetAttributesByType, GetAttributesValues } from './utils';
+import type { Attribute, Common, Utils } from '@strapi/strapi';
 
-export type BasicRelationsType = 'oneToOne' | 'oneToMany' | 'manyToOne' | 'manyToMany';
-export type PolymorphicRelationsType = 'morphToOne' | 'morphToMany' | 'morphOne' | 'morphMany';
+export type BasicRelationsType =
+  | 'oneToOne'
+  | 'oneToMany'
+  | 'manyToOne'
+  | 'manyToMany'
+  | 'morphOne'
+  | 'morphMany';
+export type PolymorphicRelationsType = 'morphToOne' | 'morphToMany';
 export type RelationsType = BasicRelationsType | PolymorphicRelationsType;
 
-export interface BasicRelationAttributeProperties<
-  S extends SchemaUID,
-  R extends RelationsType,
-  T extends SchemaUID
-> {
+export type BasicRelationProperties<
+  S extends Common.UID.Schema,
+  R extends BasicRelationsType,
+  T extends Common.UID.Schema
+> = {
   relation: R;
   target: T;
-  inversedBy?: RelationsKeysFromTo<T, S>;
-  mappedBy?: RelationsKeysFromTo<T, S>;
-}
+} & R extends `morph${string}`
+  ? {
+      morphBy?: Utils.Object.KeysBy<
+        Common.Schemas[T]['attributes'],
+        Attribute.Relation<Common.UID.Schema, Attribute.PolymorphicRelationsType>
+      >;
+    }
+  : {
+      inversedBy?: RelationsKeysFromTo<T, S>;
+      mappedBy?: RelationsKeysFromTo<T, S>;
+    };
```

Comment:
Similar thought here, can we make these into something like PolymorphicAttributes and BasicAttributes?

Reply:
I just meant any way to break it apart so the conditional can be `? PolymorphicBy : BasicBy` since these types can be verbose and not always clear at a glance what they're doing.

Reply:
I'm just afraid to add boilerplate :confused: 
We already have a lot of types with very similar names in this file and I'm not sure adding two new ones will be much clearer compared to what we have. What about adding some comments around the ternary?